### ISSUE TITLE
typing.NewType isn't supported

### ISSUE THREAD

**Issue by rhofour** [python-attrs/cattrs#94]
* cattrs version: 1.0.0
* Python version: 3.8.5
* Operating System: Fedora 32

### Description

NewType seems to break structuring / destructuring.

Here's a minimal test case:
```
import unittest
from typing import NewType
import attr
import cattr

ConfigId = NewType('ConfigId', int)

@attr.s(auto_attribs=True)
class ClassWithNewTypeMember:
    configId: ConfigId

class TestAttrs(unittest.TestCase):
    def test_newType(self):

        event = ClassWithNewTypeMember(configId = ConfigId(0))
        unstructured = cattr.unstructure(event)
        structured = cattr.structure(unstructured, ClassWithNewTypeMember)
```

I would expect this test to pass, but instead I get the following:
`E       ValueError: Unsupported type: <function NewType.<locals>.new_type at 0x7fa0349398b0>. Register a structure hook for it.
`

The workaround is to manually register a structure hook like so:
`cattr.register_structure_hook(ConfigId, lambda d, _: ConfigId(d))`

A destructuring hook isn't needed.

**Comment by letalvoj** [comment]
According to the documentation https://docs.python.org/3/library/typing.html#newtype

> The static type checker will treat the new type as if it were a subclass of the original type. This is useful in helping catch logical errors:
```
def get_user_name(user_id: UserId) -> str:
    ...

# typechecks
user_a = get_user_name(UserId(42351))

# does not typecheck; an int is not a UserId
user_b = get_user_name(-1)
```

Hence, as `ConfigId` is more specific than `int`, `cattrs` should _not_ know how to derive an encoder for it.

**Comment by Tinche** [comment]
We probably just need to teach cattrs to treat newtypes as their underlying types. I agree this would be useful.

**Comment by uskudnik** [comment]
Okay, might've found a solution by using `register_structure_hook_func`

```
import cattr
from typing import NewType

cattr.register_structure_hook_func(
    lambda t: t.__class__ is NewType, lambda d, t: t.__call__(d)
)

```

:crossed_fingers: nothing :boom: too much :slightly_smiling_face:

**Comment by raabf** [comment]
Before python 3.10, `typing.NewType` is a function which returns a function, hence isinstance does not work. In python 3.10 the symbol `typing.NewType` became an own class where you can do normal `isinstance{obj, typing.NewType)` checks (or as you noticed `t.__class__ is NewType` which does the same).

Something like `isinstance(ConfigId, type)` should not work on python 3.10 because `ConfigId` is then an instance of the `typing.NewType` class instead of an own generated function.

Heve a look at [this `is_new_type()`](https://github.com/ilevkivskyi/typing_inspect/blob/8f6aa2075ba448ab322def454137e7c59b9b302d/typing_inspect.py#L257-L279) function for a complete test regardless of python version.

**Comment by henryiii** [comment]
I'm looking for a way to register a custom converter via a type; that is, I'd like to do a little post processing on fields marked with `URL`, but I don't need a custom type for the `URL` in the classes, just `str` is fine. I could define a field for each URL, but doing it via types is simpler and cleaner, and I don't really need a converter for setting an instance (since they are frozen, no real difference). It seems like NewType would have been a nice way to do it.

From what I can tell above (which is literally asking for the opposite of what I want to do - though I'm fine if it also works by default but is customizable), what I want was supported before 3.10, but now it's broken?

**Comment by Tinche** [comment]
NewTypes are in, and will be released with the 22.2.0 release!
